### PR TITLE
chore: 16221 `swirlds-state-impl` module

### LINKED ISSUE #16221: Introduce `swirlds-state-impl` module
There will be multiple iterations of this module, but in the first iteration, it should only contain implementation classes that need to be moved from `swirlds-state-api`:
- package `com.swirlds.state.merkle`
- `MigrationContext`, `Service` and schema-related classes.

To minimize the changeset packages should remain the same unless  it makes sense to change them for one reason or the other. Test classes should be moved as well.

### PR DESCRIPTION
**Pull request by imalygin**
**Description**:

This PR introduces `swirlds-state-impl` module. This module contains classes related to the implementation of `swirlds-state-api`.

This PR migrates classes from  `swirlds-state-api` to `swirlds-state-impl`

**Related issue(s)**:

Fixes #16221 

**Notes for reviewer**:

**PLEASE DON'T BE INTIMIDATED BY THE SIZE OF THIS PR**

It does only three things:
- moves classes from one directory to another
- updates imports
- fixes module dependencies

**THERE ARE NO CHANGES IN THE CODE LOGIC** 

Just scroll it through, it shouldn't take too long. 

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by timo0 — APPROVED** [review]

**Review by mhess-swl — COMMENTED** [review]
There are a number of modules that now depend on the `swirlds-state-impl` module instead of remaining decoupled by depending on the api only. I assume this is in preparation for more changes coming?

**Comment by imalygin** [comment]
@mhess-swl 

> I assume this is in preparation for more changes coming?

Yes, there will be more changes. It's a preparation for migration of classes from `swirlds-platform-core`. Also, I briefly checked the dependencies on `swirlds-state-impl`, and all of them are justified. It's okay for the app to be dependent on the state implementation details. But it's not okay for the Block Node to be dependent on `swirlds-platform-core`, as it has a lot of unrelated functionality.

**Review by lpetrovic05 — APPROVED** [review]

**Review by mhess-swl — APPROVED** [review]

**Review by mhess-swl — DISMISSED** [review]

**Review by andrewb1269hg — DISMISSED** [review]
Reviewed the following files:
```
hedera-node/hedera-app-spi/src/main/java/module-info.java
hedera-node/hedera-app-spi/src/testFixtures/java/module-info.java
hedera-node/hedera-app/build.gradle.kts
hedera-node/hedera-app/src/main/java/module-info.java
hedera-node/hedera-app/src/testFixtures/java/module-info.java
hedera-node/hedera-schedule-service/build.gradle.kts
hedera-node/hedera-token-service/src/main/java/module-info.java
hedera-node/hedera-util-service/src/main/java/module-info.java
platform-sdk/swirlds-platform-core/build.gradle.kts
platform-sdk/swirlds-platform-core/src/main/java/module-info.java
platform-sdk/swirlds-platform-core/src/testFixtures/java/module-info.java
platform-sdk/swirlds-state-api/build.gradle.kts
platform-sdk/swirlds-state-api/src/main/java/module-info.java
platform-sdk/swirlds-state-api/src/testFixtures/java/module-info.java
platform-sdk/swirlds-state-impl/build.gradle.kts
platform-sdk/swirlds-state-impl/src/main/java/module-info.java
platform-sdk/swirlds-state-impl/src/testFixtures/java/module-info.java
```

These files look good.

**Review by Nana-EC — APPROVED** [review]

**Review by kimbor — APPROVED** [review]

**Review by tinker-michaelj — DISMISSED** [review]
LGTM, tyvm @imalygin !

**Review by andrewb1269hg — DISMISSED** [review]
Reapproved.

**Review by Nana-EC — DISMISSED** [review]

**Review by kimbor — DISMISSED** [review]

File: hedera-node/hedera-app/src/main/java/module-info.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
     requires transitive com.swirlds.metrics.api;
     requires transitive com.swirlds.platform.core;
     requires transitive com.swirlds.state.api;
+    requires transitive com.swirlds.state.impl;
```
Comment (artemananiev) [inline hedera-node/hedera-app/src/main/java/module-info.java:24]: Since it is transitive, and state.impl requires state.api, is state.api above redundant?

**Review by mxtartaglia-sl — DISMISSED** [review]

**Review by mhess-swl — DISMISSED** [review]

**Comment by artemananiev** [comment]
The changes look fine, but may be sub-optimal. `swirlds-state-api` module is minimal, it only contains classes and methods used to work with states, read and write. Another important aspect of working with states is in `swirlds-state-impl` for some reason: how to manage states, create, migrate, versioning, etc. This part is indeed not required for hypothetical "test only" implementation of state APIs, but it's still API and should be in `swirlds-state-api` module.

So my proposal is to move a few classes a little bit:
* `swirlds-state-api` module to contain all classes and methods to read from and write to states, `com.swirlds.state.spi` package. This is so called "basic API"
* classes related to state management should also be moved to `swirlds-state-api`: `Schema`, `Service`, `SchemaRegistry`, and so on. They can be placed in a separate package like `com.swirlds.state.spi.lifecycle`. Let's call it "extended API"
* `swirlds-state-impl` will contain classes that implement basic API above on top of merkle trees. Should the module be named like `swirlds-state-impl-merkle`? I am not sure
* Extended API will be implemented in `hedera-app` module. Hedera services classes will provide all service/schema definitions

@tinker-michaelj @jasperpotts What do you think about it?

File: hedera-node/hedera-app/src/main/java/module-info.java (hunk 1 above)
Comment (imalygin) [reply hedera-node/hedera-app/src/main/java/module-info.java:23]: Yes. I removed `requires transitive com.swirlds.state.api;`

**Review by mhess-swl — APPROVED** [review]

**Review by artemananiev — APPROVED** [review]

**Review by mxtartaglia-sl — APPROVED** [review]

**Review by andrewb1269hg — APPROVED** [review]
Reapproved.

**Review by lpetrovic05 — APPROVED** [review]

**Comment by tinker-michaelj** [comment]
> The changes look fine, but may be sub-optimal. `swirlds-state-api` module is minimal, it only contains classes and methods used to work with states, read and write. Another important aspect of working with states is in `swirlds-state-impl` for some reason: how to manage states, create, migrate, versioning, etc. This part is indeed not required for hypothetical "test only" implementation of state APIs, but it's still API and should be in `swirlds-state-api` module.
> 
> So my proposal is to move a few classes a little bit:
> 
> * `swirlds-state-api` module to contain all classes and methods to read from and write to states, `com.swirlds.state.spi` package. This is so called "basic API"
> * classes related to state management should also be moved to `swirlds-state-api`: `Schema`, `Service`, `SchemaRegistry`, and so on. They can be placed in a separate package like `com.swirlds.state.spi.lifecycle`. Let's call it "extended API"
> * `swirlds-state-impl` will contain classes that implement basic API above on top of merkle trees. Should the module be named like `swirlds-state-impl-merkle`? I am not sure
> * Extended API will be implemented in `hedera-app` module. Hedera services classes will provide all service/schema definitions
> 
> @tinker-michaelj @jasperpotts What do you think about it?

Makes sense to me Artem; it would be nice at some point if we could extract the common logic between the `OrderedServiceMigrator`/`MerkleSchemaRegistry` and `FakeServiceMigrator`/`FakeSchemaRegistry` types and reduce duplication between the subprocess and embedded `Hedera` state lifecycles.

**Comment by imalygin** [comment]
>it would be nice at some point if we could extract the common logic between the OrderedServiceMigrator/MerkleSchemaRegistry and FakeServiceMigrator/FakeSchemaRegistry types and reduce duplication between the subprocess and embedded Hedera state lifecycles

Thank you for the hint, Michael. I'll create follow up ticket.

**Review by tinker-michaelj — APPROVED** [review]
LGTM, ty @imalygin !

**Comment by rbarkerSL** [comment]
@imalygin this PR seems to have broken the `timing-sensitive` tests in hedera services

See the associated run here: https://github.com/hashgraph/hedera-services/actions/runs/11709318923

**Pull request merged**